libbpf-cargo/src/test.rs

```diff
@@ -769,8 +769,94 @@ fn test_skeleton_builder_clang_opts() {
         .unwrap();
 }
 
-#[test]
-fn test_btf_dump_basic() {
+// -- TEST RUST GENERATION OF BTF PROGRAMS --
+
+/// Searches the Btf struct for a BtfType
+/// returns type identifier <u32> if found
+/// fails calling test if not found, or if duplicates
+///
+/// usage: -- search for basic struct/union/enum with exact match to name
+///        assert_type!(<Btf to search in>, <BtfType to search for>, <&str search name>);
+///        eg:
+///        let my_type = assert_type!(btf, Struct, "name")
+///    or: -- search for basic struct/union/enum/var containing name
+///        assert_type!(<Btf to search in>, <BtfType to search for>, <&str search name>, true);
+///        eg:
+///        let my_type = assert_type!(btf, Datasec, "bss", true);
+///    or: -- search for a Variable name in a Datasec -- exact match
+///        assert_type!(<Btf to search in>, Var, <&str search name>);)
+///        eg
+///        let let my_type = assert_type!(btf, Var, "name")
+macro_rules! assert_type {
+
+    // match for a named BtfType::Var inside all vars in a Datasec
+    ($btf:ident, Var, $name:literal) => {{
+        let mut asserted_type: Option<u32> = None;
+        for ty in $btf.types().iter() {
+            match ty {
+                btf::BtfType::Datasec(t) => {
+                    for var in &t.vars {
+                        let var_ty = $btf.type_by_id(var.type_id)
+                            .expect("Failed to lookup datasec var");
+                        match var_ty {
+                            btf::BtfType::Var(t) => {
+                                if t.name == $name {
+                                    assert!(asserted_type.is_none());
+                                    asserted_type = Some(var.type_id);
+                                }
+                            }
+                            _ => panic!("Datasec var didn't point to a var. Instead: {}", var_ty),
+                        }
+                    }
+                }
+                _ => (),
+            }
+        }
+        assert!(asserted_type.is_some());
+        asserted_type.unwrap()
+    }};
+
+    // match for a named BtfType.  If contains == true then do not match for exact name
+    ($btf:ident, $btf_type:ident, $name:literal $(, $contains:expr)?) => {{
```

Comment:
An optional macro param is a little complicated to read. Is it possible to make this a 4-arg macro and then have 2 other 3-arg macros, `assert_type` and `assert_type_substring`, wrap the 4-arg macro?

Reply:
While not exact to what you've requested, I think I have simplified the "optional" style macro style you disliked.

If your intention was that the test-writer should use find_type_in_btf vs find_type_in_btf_substring let me know.